中文文档/波场协议/多重签名.md

```diff
@@ -186,21 +186,140 @@ Active权限有以下特性：\
 2、修改permission\
 3、创建合约，签名\
 4、发送交易
+ 
+**http-demo**
+```
+http://{{host}}:{{port}}/wallet/accountpermissionupdate
+
+
+{
+  "owner_address": "41ffa9466d5bf6bb6b7e4ab6ef2b1cb9f1f41f9700",
+  "owner": {
+    "type": 0,
+    "permission_name": "owner",
+    "threshold": 2,
+    "keys": [{
+        "address": "41F08012B4881C320EB40B80F1228731898824E09D",
+        "weight": 1
+      },
+      {
+        "address": "41DF309FEF25B311E7895562BD9E11AAB2A58816D2",
+        "weight": 1
+      },
+      {
+        "address": "41BB7322198D273E39B940A5A4C955CB7199A0CDEE",
+        "weight": 1
+      }
+    ]
+  },
+  "witness": {
+      "type": 1,
+      "permission_name": "witness",
+      "threshold": 1,
+      "keys": [{
+          "address": "41F08012B4881C320EB40B80F1228731898824E09D",
+          "weight": 1
+        } 
+      ]
+    },
+  "actives": [{
+    "type": 2,
+    "permission_name": "active0",
+    "threshold": 3,
+    "operations": "7fff1fc0037e0000000000000000000000000000000000000000000000000000",
+    "keys": [{
+        "address": "41F08012B4881C320EB40B80F1228731898824E09D",
+        "weight": 1
+      },
+      {
+        "address": "41DF309FEF25B311E7895562BD9E11AAB2A58816D2",
+        "weight": 1
+      },
+      {
+        "address": "41BB7322198D273E39B940A5A4C955CB7199A0CDEE",
+        "weight": 1
+      }
+    ]
+  }]
+}
+
+参数字段的定义及限制，请查看"2.1 结构说明"。
+
+```
+
+**active权限中operations的计算示例**
+
+``` java
+public static void main(String[] args) {
+
+  //指定需要支持的合约id(查看proto中Transaction.ContractType定义)，这里包含除AccountPermissionUpdateContract（id=46）以外的所有合约  
+  Integer[] contractId = {0, 1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 30, 31,
+      32, 33, 41, 42, 43, 44, 45};
+  List<Integer> list = new ArrayList<>(Arrays.asList(contractId)); 
+  byte[] operations = new byte[32];
+  list.forEach(e -> {
+    operations[e / 8] |= (1 << e % 8);
+  });
+  
+  //7fff1fc0037e0000000000000000000000000000000000000000000000000000
+  System.out.println(ByteArray.toHexString(operations));
+}
+```
+
 
 ### <h2 id="3.2">3.2 执行合约</h2> 
 
-1、创建交易\
-2、用户A签名，将签名后交易通过其他方式发送给B。\
-3、用户B签名，将签名后交易通过其他方式发送给C。\
+1、创建交易，与非多重签名交易的构建过程相同\
+2、指定Permission_id，默认为0，表示owner-permission, demo见https://github.com/tronprotocol/wallet-cli/commit/ff9122f2236f1ce19cbb9ba9f0494c8923a3d10c#diff-a63fa7484f62fe1d8fb27276c991a4e3R211 \
+3、用户A签名，将签名后交易通过其他方式发送给B。\
+4、用户B签名，将签名后交易通过其他方式发送给C。\
 …\
 n、最后一个完成签名的用户，将交易广播到节点。\
 n+1、验证多重签名的权重之和大于域值则接受交易，否则拒绝交易
 
 代码示例：
 
 https://github.com/tronprotocol/wallet-cli/blob/multi_sign_V2/src/main/java/org/tron/demo/MultiSignDemo.java
+  
 
-## <h2 id="4">4.其他问题</h2>  
-1、兼容性说明  
+### <h2 id="3.3">3.3 其他新增接口</h2>
+接口详细说明，请查看Tron-http.md与波场钱包RPC-API.md \
+1、增加签名
+```
+curl -X POST  http://127.0.0.1:8090/wallet/addtransactionsign -d '{"transaction": "TransferContract", "privateKey": "permissionkey1"}'
```

Comment:
http://127.0.0.1:8090/wallet/addtransactionsign

Reply:
What is the problem?